include/ldsCtrlEst_h/lds_mpcctrl.h

```diff
@@ -0,0 +1,533 @@
+OSQPSolution* MpcController<System>::slow_update(const Vector& x0,
+                                                 const Vector& u0,
+                                                 const Matrix& xr,
+                                                 size_t n_sim) {
+  Matrix leq = join_horiz(
+      -x0.t(), arma::zeros((N_ - 1) * n_).t());  // Lower equality bound
+  Matrix ueq = leq;                              // Upper equality bound
+
+  // Update x over n_sim many steps
+  Matrix Axs = arma::real(
+      arma::powmat(A_, static_cast<double>(n_sim)));  // State multiplier
+  Matrix Aus = arma::zeros(n_, n_);                   // Input multiplier
+  for (int i = 0; i < n_sim; i++) {
+    Aus += arma::powmat(A_, i);
+  }
+
+  // Ax + Bu = 0
+  Matrix Ax(arma::kron(arma::speye<arma::SpMat<data_t>>(N_, N_),
+                      -arma::speye<arma::SpMat<data_t>>(n_, n_)) +
+            arma::kron(eye_offset(N_), arma::SpMat<data_t>(Axs)));
+  Matrix B0(1, M_);
+  Matrix Bstep(M_, M_, arma::fill::eye);
+  Matrix Bend = arma::join_horiz(Matrix(N_ - M_ - 1, M_ - 1),
+                                 Matrix(N_ - M_ - 1, 1, arma::fill::ones));
+  Matrix Bu = kron(join_vert(B0, Matrix(Bstep), Bend), Aus * B_);
+  Matrix Aeq = join_horiz(Ax, Bu);  // Equality condition
+
+  Acon_ = join_vert(Aeq, Matrix(Aineq_));  // Update condition
+  lb_ = join_horiz(leq, lineq_).t();       // Lower bound
+  ub_ = join_horiz(ueq, uineq_).t();       // Upper bound
+
+  // Convert state penalty from reference to OSQP format
+  Vector q_arma;
+  {
+    arma::uvec indices = arma::regspace<arma::uvec>(0, n_sim, N_ * n_sim - 1);
+    Matrix sliced_xr = xr.cols(indices);
+    Matrix Qxr_full = -2 * Q_ * sliced_xr;
+    Vector Qxr = Qxr_full.as_col();  // Qxr for every simulation time step
+
+    Vector qu =
+        join_vert((-2 * S_ * u0), Vector((M_ - 1) * m_, arma::fill::zeros));
+    Vector qx = Qxr.rows(0, N_ * n_ - 1);
+    q_arma = join_vert(qx, qu);
+
+  }
+
+  // Problem values for OSQP
+  OSQPFloat* q = from_vec(q_arma);
+  OSQPFloat* lb = from_vec(lb_);
+  OSQPFloat* ub = from_vec(ub_);
+
+  // Solver, settings, matrices
+  OSQPSolver* solver;
+  OSQPSettings* settings;
+  OSQPCscMatrix* A = from_matrix(Acon_);
+  OSQPCscMatrix* P = from_matrix(P_);
+
+  // Set settings
+  settings = (OSQPSettings*)malloc(sizeof(OSQPSettings));
+  if (settings) {
+    osqp_set_default_settings(settings);
+    settings->verbose = false;
+    settings->warm_starting = false;
+  }
+
+  // Set up solver
+  OSQPInt exit_flag = osqp_setup(&solver, P, q, A, lb, ub, A->m, A->n, settings); // n and m are set by A cols and rows
+
+  // Solve problem
+  if (!exit_flag) exit_flag = osqp_solve(solver);
+
+  if (solver->info->status_val != OSQP_SOLVED) {
+    std::string error_message = "OSQP solver failed to solve.";
+    throw std::runtime_error(error_message);
+  }
+
+  OSQPSolution* sol = std::move(solver->solution);
+
+  // Clean up
+  osqp_cleanup(solver);
+  if (q) free(q);
```

Comment:
why free() instead of delete? (not saying delete is right; I'm not an expert on the finer points of memory management)

Reply:
As far as I know, `free()` is for freeing memory allocated with `malloc`, whereas `delete` is used for freeing memory allocated with `new`, but I will verify that to make sure I'm not doing anything wrong.

Reply:
The follow-up question then is why `malloc` instead of `new`?

Reply:
I don't actually have a good reason for using either. I mostly just followed what the other controllers did and tried to align closely to that.